Group Undos

disordinary opened the pull request:
Refs #502

Basically there are three types of undo events, `insert`, `delete`, and everything else (which is undefined here).

`Insert` and `delete` events group together overwriting the last element in the undo queue when they are repeated unless another event occurs which breaks it or a timeout occurs.

So, if I write text then as long as I don't delete text or insert an atom or card, and as long as the timeout doesn't occur, those events are grouped by replacing the last element on the undo queue and are popped in one go.

A timeout is reset whenever the `run` method is called on the editor, so it doesn't happen on the first occurrence of an event in a run but rather the last.

bantic commented on src/js/editor/editor.js:
```diff
@@ -70,6 +70,16 @@ const CALLBACK_QUEUES = {
+  DELETE: 2
+};
+
+const UNDO_RUN_BREAK_TIMEOUT_LENGTH = 500; // how many ms after the last event before an UNDO run is broken
```
Mentioned in gitter: Increasing this seems worthwhile. 5000 ms is a good default. I'm wondering if it makes sense to allow it to be customizable, also...

bantic commented on src/js/editor/editor.js:
```diff
@@ -293,7 +306,7 @@ class Editor {
       let nextPosition = postEditor.deleteAtPosition(position, direction, {unit});
       postEditor.setRange(nextPosition);
-    });
+    }, EDITOR_RUN_LAST_ACTION_STATE.DELETE);
```
I am reluctant to modify the method signature of the already-public `run` method in order to add a private-ish second parameter...

bantic replied:
Also, there isn't necessarily anything about the `run` method that requires it to be only one of ["insert","delete"] — it's totally possible to do an `editor.run` that inserts at some places and deletes at others.
Not sure how important this distinction is. The most common use case is certainly for a `run` to focus on either one or the other.

disordinary replied:
I think if it's mixed then it's unlikely to be the result of a direct manipulation by the end user and therefore should break the undo grouping. The same as some events which are insertions like a Paste event.

The reason that this works is because only two states are explicitly defined (TEXT_INSERT and DELETE which results in a grouping) and everything else is undefined which keeps them as unique elements on the undo stack.

disordinary replied:
Yes, I agree with the private argument on run. Alternatives include:
- Having a more complicated return type that can include the undo grouping information.
- Duplicating Run which focuses on those two specific use cases.

bantic replied:
@disordinary Sorry to let this linger — we were at EmberFest.
Yeah, hiding the second arg with a private `_run` totally makes sense to me. 👍

bantic commented on src/js/editor/editor.js:
```diff
@@ -673,10 +686,32 @@ class Editor {
+          this._undoRunTimeout = null;
+          this._undoRunLastState = null;
+        },
+        UNDO_RUN_BREAK_TIMEOUT_LENGTH);
```
`this._undoRunTimeout` probably needs to be explicitly cleared in `editor#destroy` as well (if it is present)

bantic commented on tests/acceptance/editor-undo-redo-test.js:
```diff
@@ -99,6 +92,46 @@ test('undo/redo the insertion of multiple characters', (assert) => {
   });
 });
 
+test('adding and deleting characters break the undo group/run', (assert) => {
```
Adding a test for the timeout functionality would also be good here (and is another reason that it could/should be configurable — at least for tests).

disordinary replied:
Yes,  or in the `undo/redo the insertion of multiple characters`  test above it.

Making the timeout configurable is a good option although I don't see people changing it too often if we come up with a timeout that works well.

bantic commented:
@disordinary Thanks again for your work on this. If you want to add the private `_run` as described and a test that exercises the timeout functionality (which I imagine necessitates making the timeout configurable), I would be happy to merge!

disordinary commented:
@bantic, great I'll get a PR together by Monday.

disordinary commented:
I've updated the PR as follows:
- Introduced a new private `_run` command.
- Set the timeout for the undo blocks as an option `option.undoBlockTimeout` with a default of 5000 seconds (feel free to rename, you know that quote about the two hard things in computer science).
- Introduced new tests for the undo grouping timeout.

disordinary commented:
@bantic, I just noticed the failing tests. It seems to be a problem with the browsers themselves not being able to connect to the test environment and not anything to do with this PR, do you mind running again?

mixonic commented:
@disordinary I restarted them. Force pushing a no-op rebase would also kick them off. 👍

mixonic commented:
@disordinary I'll look closely in just a few hours here.

mixonic commented on src/js/editor/editor.js:
```diff
@@ -677,6 +695,45 @@ class Editor {
+    }
+
+    if(isInARun) {
+        this._undoRunTimeout = setTimeout(() => {
```
This debouncing pattern isn't free- setting and clearing timeouts can be very slow JS code. Better might be to schedule a flush of this state in the future, and at the time of the flush check for if there have been any more recent actions that mean the flush should be cancelled. Then, if the flush is cancelled, schedule a new followup. This would be move the timeout setup/cancel work outside of the input and interactive parts of the flow.

mixonic commented on src/js/editor/editor.js:
```diff
@@ -677,6 +695,45 @@ class Editor {
+   * @return {Mixed} The return value of `callback`.
+   * @private
+   */
+  _run(callback, actionType) {
```
Does it feel wrong that the actionType stuff happens at this level? Wouldn't it be better to just have `deleteRange`, for example, set the state of lastState and schedule a some further work into the loop? That would seem to generically handle a lot of manual flagging of runloops, and to be more friendly to user-space code.

disordinary replied:
The reason that the state is being passed as an argument into the `run` method and not being set on the object is that `run` somehow needs to know that the command that it's executing is the command that set the state, because any event other than `insertText`, `deleteAtPosition`, and `deleteRange` should default to breaking the undo grouping.  I can't think of an elegant solution off the top of my head unfortunately.

mixonic commented:
Excellent, after some pairing today this feels much improved and we're ready to land it pending a green build. Thanks @disordinary 👏 !

mixonic commented:
I'm merging this with a failing build. IE11 was failing previously due to the racy nature of the undo/redo tests (ripe for refactoring). By expanding the delay used I've gotten them passing locally. At this point the failing build is just Sauce not having the IE environment available to run against. Gonna move it into master and get it integrated with https://github.com/bustlelabs/mobiledoc-kit/pull/515 so we can look at a new point release.
